### PR TITLE
(closes #2254) Make PSyAD harness use logging in LFRic version

### LINKED ISSUE #2254: [LFRic] PSyAD generated adjoint tests do not throw error if adjoint test fails
Current implementation of adjoint tests would mean that a failing adjoint test(s) would not fail the automated testing of the psyad test harness miniapp. A simple solution to this would be to include an exception thrown when relative_diff < overall_tolerance.

An example pseudo-code solution is shown below:

```
    use log_mod, only : log_event, LOG_LEVEL_ERROR, LOG_LEVEL_INFO
    ...
    ! Test the inner-product values for equality, allowing for the precision of the active variables
    MachineTol = SPACING(MAX(ABS(inner1), ABS(inner2)))
    relative_diff = ABS(inner1 - inner2) / MachineTol
    if (relative_diff < overall_tolerance) then
      !WRITE(*, *) 'Test of adjoint of ''tl_{tl_name}_kernel_type'' PASSED: ', inner1, inner2, relative_diff
      call log_event( "Test of adjoint of ''tl_{tl_name}_kernel_type'' PASSED: {inner1}, {inner2}, {relative_diff}", LOG_LEVEL_INFO )
    else
      !WRITE(*, *) 'Test of adjoint of ''tl_{tl_name}_kernel_type'' FAILED: ', inner1, inner2, relative_diff
      call log_event( "Test of adjoint of ''tl_{tl_name}_kernel_type'' FAILED: {inner1}, {inner2}, {relative_diff}", LOG_LEVEL_ERROR )
    end if
```

### PR DESCRIPTION
**Pull request by rupertford**
To fit in with the LFRic testing environment, the LFRic PSyAD harness should use LFRic logging to indicate success or failure of the LFRic adjoint code. @DrTVockerodtMO has kindly provided the required code in the associated issue #2254.

### CODE REVIEW

**Review by arporter — CHANGES_REQUESTED** [review]
Thanks for improving this. The change is basically OK but I baulked at the duplication. Therefore I'm requesting that you just specialise the part of the code that does the error reporting.
I'll check for pylint etc. next time.

File: src/psyclone/psyad/domain/lfric/lfric_adjoint_harness.py [hunk 1]
```diff
@@ -432,6 +437,85 @@ def _validate_geom_arg(kern, arg_idx, name, valid_spaces, vec_len):
...
+    :type sym_table: :py:class:`psyclone.psyir.symbols.SymbolTable`
+    :param kernel: the routine for which this adjoint test is being performed.
+    :type kernel: :py:class:`psyclone.psyir.nodes.Routine`
+    :param var1: the symbol holding the first value for the comparison.
+    :type var1: :py:class:`psyclone.psyir.symbols.DataSymbol`
+    :param var2: the symbol holding the second value for the comparison.
+    :type var2: :py:class:`psyclone.psyir.symbols.DataSymbol`
+
+    :returns: the PSyIR nodes that perform the check.
+    :rtype: list of :py:class:`psyclone.psyir.nodes.Node`
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyad/domain/lfric/lfric_adjoint_harness.py]: Modified to `List[...]` in the appropriate places

File: src/psyclone/psyad/domain/lfric/lfric_adjoint_harness.py [hunk 2]
```diff
@@ -432,6 +437,85 @@ def _validate_geom_arg(kern, arg_idx, name, valid_spaces, vec_len):
...
+    statements[-1].preceding_comment = (
+        "Test the inner-product values for equality, allowing for the "
+        "precision of the active variables")
+    sub_op = BinaryOperation.create(BinaryOperation.Operator.SUB,
+                                    Reference(var1), Reference(var2))
+    abs_op = IntrinsicCall.create(IntrinsicCall.Intrinsic.ABS, [sub_op])
+    div_op = BinaryOperation.create(BinaryOperation.Operator.DIV,
+                                    abs_op, Reference(mtol))
+    statements.append(Assignment.create(Reference(rel_diff), div_op))
+
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyad/domain/lfric/lfric_adjoint_harness.py:470]: Removed duplication as requested.

**Comment by rupertford** [comment]
Ready for next review from @arporter

**Review by arporter — COMMENTED** [review]
Requested changes have been made.
Changes are pylint/pycodestyle clean.
I've taken the liberty of fixing an error found when generating the Ref. Guide (not related to this PR).
Just waiting on the integration tests now.

**Review by arporter — APPROVED** [review]
NEMO integration tests failed but that's not the fault of this PR.
Will proceed to merge.

**Pull request merged**